### PR TITLE
Retry failures after backing off first

### PR DESCRIPTION
**Pull request by abless**

### CODE REVIEW

**Comment by justinuang** [comment]
@iamdanfox @uschi2000 we need this to release SMM 2.0 HA.

**Comment by iamdanfox** [comment]
@abless if you want to get this pushed through after-hours in London, can you coordinate with @qinfchen who is based in NY?

File: http-clients/src/main/java/com/palantir/remoting3/clients/ClientConfigurations.java [hunk 1]
```diff
@@ -40,7 +40,7 @@
     private static final Duration DEFAULT_CONNECT_TIMEOUT = Duration.ofSeconds(10);
     private static final Duration DEFAULT_READ_TIMEOUT = Duration.ofMinutes(10);
     private static final Duration DEFAULT_WRITE_TIMEOUT = Duration.ofMinutes(10);
-    private static final Duration DEFAULT_BACKOFF_SLOT_SIZE = Duration.ofMillis(500);
+    private static final Duration DEFAULT_BACKOFF_SLOT_SIZE = Duration.ofMillis(250);
```
Comment (uschi2000) [inline http-clients/src/main/java/com/palantir/remoting3/clients/ClientConfigurations.java:43]: why this change?

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java [hunk 2]
```diff
@@ -48,6 +48,6 @@
         }
 
         int upperBound = (int) Math.pow(2, retryNumber);
-        return Optional.of(backoffSlotSize.multipliedBy(random.nextInt(upperBound)));
+        return Optional.of(Duration.ofNanos(Math.round(backoffSlotSize.toNanos() * random.nextDouble() * upperBound)));
```
Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: why this change?

File: http-clients/src/main/java/com/palantir/remoting3/clients/ClientConfigurations.java (hunk 1 above)
Comment (abless) [reply http-clients/src/main/java/com/palantir/remoting3/clients/ClientConfigurations.java:43]: This PR changes in behavior in that we now always apply a backoff duration when calls encounter IOExceptions (e.g. ConnectExceptions), whereas previous implementation would retry right away. When load-testing rolling upgrades with one of our services we noticed that this would mean that one server (the only one alive) would get hit with lots of requests all at the same time, causing ConnectExceptions.

We are halving the default backoff slot size because the previous implementation of ExponentialBackoff didn't actually adhere to its javadoc - it says it chooses a backoff from the interval of [0, backoffSlotSize * 2^c) but it really chose it from [0, backoffSlotSize * 2^(c-1)) - this has been fixed in the implementation so we're halving the default to provide consistent backoff duration.

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java (hunk 2 above)
Comment (abless) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: To increase entropy. random.nextInt() has pretty low variance for small number of retries (in particular, for the first retry) and so to make sure that lots of calls don't retry all at the same time we increase the entropy/variance by going more granular.

Comment (uschi2000) [inline okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: the Math.round seems redundant

File: okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/ExponentialBackoffTest.java [hunk 3]
```diff
@@ -45,17 +42,14 @@ public void testRetriesCorrectNumberOfTimesAndFindsRandomBackoffWithInExponentia
...
         assertThat(backoff.nextBackoff()).contains(ONE_SECOND.multipliedBy(2));
-        verify(random).nextInt(4);
 
-        when(random.nextInt(anyInt())).thenReturn(3);
-        assertThat(backoff.nextBackoff()).contains(ONE_SECOND.multipliedBy(3));
-        verify(random).nextInt(8);
+        when(random.nextDouble()).thenReturn(1.0);
+        assertThat(backoff.nextBackoff()).contains(ONE_SECOND.multipliedBy(4));
+
+        when(random.nextDouble()).thenReturn(1.0);
```
Comment (uschi2000) [inline okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/ExponentialBackoffTest.java]: let's vary the random number returned.

**Review by uschi2000 — COMMENTED** [review]
ok, sounds good. two nitpicks, otherwise +1

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java (hunk 2 above)
Comment (abless) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: Duration.ofNanos() expects a long, since we're using `random.nextDouble()` we need to round.

File: okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/ExponentialBackoffTest.java (hunk 3 above)
Comment (abless) [reply okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/ExponentialBackoffTest.java]: Can do, but then it's not as obvious that the backoff is exponential (2, 4, 8) - still want to vary it?

Comment (uschi2000) [reply okhttp-clients/src/test/java/com/palantir/remoting3/okhttp/ExponentialBackoffTest.java]: yep. just change the last one to 0.5 --> 4.

File: okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java (hunk 2 above)
Comment (uschi2000) [reply okhttp-clients/src/main/java/com/palantir/remoting3/okhttp/ExponentialBackoff.java:51]: I was thinking to just do `(long) d`, but sure, round also works.

**Review by uschi2000 — APPROVED** [review]

**Comment by justinuang** [comment]
Can we get a release too if it's not done yet

**Comment by gracew** [comment]
tagged: https://github.com/palantir/http-remoting/releases/tag/3.20.0
